trl/trainer/sft_trainer.py

```diff
@@ -974,11 +975,22 @@ def tokenize_fn(example, processing_class, dataset_text_field, assistant_only_lo
                                 "token handling. Verify that the tokenizer is processing text consistently."
                             )
 
-                        # Create a completion mask
-                        completion_mask = [0] * len(prompt_ids) + [1] * (len(prompt_completion_ids) - len(prompt_ids))
                         output["input_ids"] = prompt_completion_ids
+
+                        # Create completion_mask - more precise thanks to add_generation_prompt=True
+                        completion_mask = [0] * len(prompt_ids) + [1] * (len(prompt_completion_ids) - len(prompt_ids))
                         output["completion_mask"] = completion_mask
 
+                        # Warn if assistant_masks are all zeros (chat template doesn't support {% generation %})
+                        # This means assistant_only_loss won't work correctly with prompt-completion format
+                        if "assistant_masks" in output and 1 not in output["assistant_masks"] and assistant_only_loss:
+                            logger.warning(
+                                "You're using `assistant_only_loss=True` with prompt-completion format, but the chat template "
+                                "doesn't generate valid assistant masks — it may be missing the `{% generation %}` keyword. "
+                                "This will cause all tokens to be masked during training. Please use `completion_only_loss=True` "
+                                "instead to train on completion tokens."
```

Comment:
```suggestion
                                "This will cause all tokens to be masked during training."
```

no, I think you're mixing completion-only and assistant-only. See https://huggingface.co/docs/trl/en/sft_trainer#train-on-assistant-messages-only and https://huggingface.co/docs/trl/en/sft_trainer#train-on-completion-only